src/operator/sum.rs

```diff
@@ -0,0 +1,201 @@
+//! N-ary plus operator.
+
+use crate::{
+    algebra::{AddAssignByRef, AddByRef, HasZero, WithNumEntries},
+    circuit::{
+        operator_traits::{NaryOperator, Operator},
+        Circuit, OwnershipPreference, Scope, Stream,
+    },
+};
+use std::{borrow::Cow, iter::once, mem::transmute, ops::Add};
+
+impl<P, D> Stream<Circuit<P>, D>
+where
+    P: Clone + 'static,
+    D: Add<Output = D> + AddByRef + AddAssignByRef + Clone + HasZero + WithNumEntries + 'static,
+{
+    /// Apply the [`Sum`] operator to `self` and all streams in `streams`.
+    pub fn sum<'a, I>(&'a self, streams: I) -> Stream<Circuit<P>, D>
+    where
+        I: Iterator<Item = &'a Self>,
+    {
+        self.circuit()
+            .add_nary_operator(Sum::new(), once(self).chain(streams))
+    }
+}
+
+/// Operator that computes the sum of values across all its input streams at
+/// each timestamp.
+pub struct Sum<D>
+where
+    D: Clone + 'static,
+{
+    // Vector of input values.  Keep it here to reuse the allocation across
+    // operator invocations.
+    inputs: Vec<Cow<'static, D>>,
+}
+
+impl<D> Sum<D>
+where
+    D: Clone + 'static,
+{
+    pub fn new() -> Self {
+        Self { inputs: Vec::new() }
+    }
+}
+
+impl<D> Default for Sum<D>
+where
+    D: Clone + 'static,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<D> Operator for Sum<D>
+where
+    D: Clone + 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Sum")
+    }
+
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<D> NaryOperator<D, D> for Sum<D>
+where
+    D: Add<Output = D> + AddAssignByRef + Clone + HasZero + WithNumEntries + 'static,
+{
+    fn eval<'a, Iter>(&mut self, inputs: Iter) -> D
+    where
+        Iter: Iterator<Item = Cow<'a, D>>,
+    {
+        let input_vec = unsafe {
+            transmute::<&mut Vec<Cow<'static, D>>, &mut Vec<Cow<'a, D>>>(&mut self.inputs)
+        };
```

Comment:
I'm fairly sure this is UB, you probably want to do something like this:
```suggestion
        let input_vec = unsafe {
            let buffer = ManuallyDrop::new(take(&mut self.inputs));
            let (ptr, len, cap) = (buffer.as_mut_ptr(), buffer.len(), buffer.capacity());

            Vec::from_raw_parts(ptr, len, cap)
        };
```
and then the inverse at the end of the function to replace `self.inputs`

Reply:
Why is your alternative safer? It still changes the lifetime while the compiler isn't looking.

Reply:
It's how it does it, funky stuff to do with variance. Additionally, my code actually takes ownership of the vector instead of mutably aliasing it while another mutable reference is alive to the struct containing it (meaning `self.inputs` is implicitly borrowed via the borrow on self) which avoids the good ole double mut refs problem

Reply:
Ok, thanks, let me try this.